File: C/15_DP1/BOJ_17404_RGB거리2.c
#include <stdio.h>
//#define math_min(a, b) a < b ? a : b
int dp[1000][3], cost[1000][3];
int math_min(int a, int b) {
	return a < b ? a : b;
}
main() {
	int n;
	scanf("%d", &n);
	for (int i = 0; i < n; i++) {
		scanf("%d %d %d", &cost[i][0], &cost[i][1], &cost[i][2]);
	}
	int result = 2147483647;
	for (int i = 0; i < 3; i++) {
		for (int j = 0; j < 3; j++) {
			if (i == j) {
				dp[0][j] = cost[0][i];
			}
			else {
				dp[0][j] = 1000 * 1000 + 1;
			}
		}
		for (int j = 1; j < n; j++) {
			dp[j][0] = math_min(dp[j - 1][1], dp[j - 1][2]) + cost[j][0];
			dp[j][1] = math_min(dp[j - 1][0], dp[j - 1][2]) + cost[j][1];
			dp[j][2] = math_min(dp[j - 1][0], dp[j - 1][1]) + cost[j][2];
		}
		for (int j = 0; j < 3; j++) {
			if (j != i) {
				result = math_min(result, dp[n - 1][j]);
			}
		}
	}
	printf("%d\n", result);
}

File: JAVA/15_DP1/BOJ_17404_RGB거리2.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    final static int red = 0;
    final static int green = 1;
    final static int blue = 2;

    static int[][] cost;
    static int[][] dp;

    public static int paint(int n, int color, int fix) {
        if (n == 0) {
            if (color == fix) {
                dp[n][color] = 1000001;
            } else {
                dp[n][color] = cost[n][color];
            }
        }
        if (dp[n][color] == 0) {
            if (color == 0) {
                dp[n][0] = Math.min(paint(n - 1, 1, fix), paint(n - 1, 2, fix)) + cost[n][0];
            } else if (color == 1) {
                dp[n][1] = Math.min(paint(n - 1, 0, fix), paint(n - 1, 2, fix)) + cost[n][1];
            } else {
                dp[n][2] = Math.min(paint(n - 1, 0, fix), paint(n - 1, 1, fix)) + cost[n][2];
            }
        }
        return dp[n][color];
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        cost = new int[n][3];
        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine(), " ");
            cost[i][0] = Integer.parseInt(st.nextToken());
            cost[i][1] = Integer.parseInt(st.nextToken());
            cost[i][2] = Integer.parseInt(st.nextToken());
        }
        int result = Integer.MAX_VALUE;
        for (int i = 0; i < 3; i++) {
            dp = new int[n][3];
            result = Math.min(result, paint(n - 1, i, i));
        }
        System.out.println(result);
    }
}